### PR TITLE
Add placeholder support

### PR DESCRIPTION
**Pull request by trescube**
Call placeholder service when libpostal interprets input to be only admin areas (no house number, street, query, or category).  

Skips further processing if placeholder returned results but falls back to existing production behavior otherwise.  

Disables placeholder support if `api.services.placeholder.url` is not defined in pelias.json.  

Fixes pelias/pelias#548

### CODE REVIEW

File: service/placeholder.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                catch (err) {
+                  logger.error(`${encodeURI(requestUrl)} could not parse response: ${data}`);
+                  return callback(`${encodeURI(requestUrl)} could not parse response: ${data}`);
+                }
+              }
+              else {
+                // otherwise there was a non-200 status so handle generically
+                logger.error(`${encodeURI(requestUrl)} returned status ${response.statusCode}: ${data}`);
+                return callback(`${encodeURI(requestUrl)} returned status ${response.statusCode}: ${data}`);
+              }
```
Comment (orangejulius) [inline service/placeholder.js]: Great error handling here. Differentiating between status code and parsing issues will help track down any issues

File: service/placeholder.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  }
+
+  logger.info(`using placeholder service at ${url}`);
+  return {
+    search: (text, lang, callback) => {
+      const requestUrl = `${url}/search?text=${text}&lang=${lang}`;
+
+      request
+        .get(requestUrl)
```
Comment (orangejulius) [inline service/placeholder.js]: wouldn't something like this work? https://github.com/pelias/api/blob/master/service/pointinpolygon.js#L12

or does that actually only work if the response is a certain size or smaller?

File: controller/predicates/is_admin_only_analysis.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const _ = require('lodash');
+
+module.exports = (request, response) => {
+  if (!request.clean.hasOwnProperty('parsed_text')) {
+    return false;
+  }
+
+  // return true only if all non-admin properties of parsed_text are empty
+  return ['number', 'street', 'query', 'category'].every((prop) => {
```
Comment (dianashk) [inline controller/predicates/is_admin_only_analysis.js]: should this include `house`?

Comment (trescube) [inline controller/predicates/is_admin_only_analysis.js]: No, this is dealing with the translated field names, so `house` in libpostal becomes `query` in the API.

File: controller/placeholder.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const _ = require('lodash');
+
+const logger = require('pelias-logger').get('api');
+const logging = require( '../helper/logging' );
+const Document = require('pelias-model').Document;
+
+function synthesizeDocs(result) {
+  return result.lineage.map((hierarchy) => {
```
Comment (missinglink) [inline controller/placeholder.js]: it's possible that some WOF records have no hierarchy and so might have no lineage (as in not even an empty array):

- https://github.com/pelias/placeholder/blob/master/prototype/wof.js#L23
- https://github.com/whosonfirst-data/whosonfirst-data/issues/739

File: routes/v1.js [hunk 5]
```diff
@@ -79,13 +81,20 @@ const hasResponseDataOrRequestErrors = any(hasResponseData, hasRequestErrors);
 function addRoutes(app, peliasConfig) {
   const esclient = elasticsearch.Client(peliasConfig.esclient);
 
-  const isPipServiceEnabled = require('../controller/predicates/is_pip_service_enabled')(peliasConfig.api.pipService);
+  const isPipServiceEnabled = require('../controller/predicates/is_service_enabled')(peliasConfig.api.pipService);
+  const isPlaceholderServiceEnabled = require('../controller/predicates/is_service_enabled')(peliasConfig.api.placeholderService);
+
   const pipService = require('../service/pointinpolygon')(peliasConfig.api.pipService);
+  const placeholderService = require('../service/placeholder')('http://localhost:3000/parser');
```
Comment (missinglink) [inline routes/v1.js]: localhost:3000?

**Comment by missinglink** [comment]
no code comments?

File: routes/v1.js (hunk 5 above)
Comment (trescube) [inline routes/v1.js]: D'oh!  Inadvertent hardcoding.

File: controller/placeholder.js (hunk 4 above)
Comment (trescube) [inline controller/placeholder.js]: That's probably the problem we saw yesterday with the acceptance-tests.

**Comment by trescube** [comment]
Adds support for general HTTP/JSON service with service configs configurable in `api` section: 

```json
  "api": {
    "legacyUrl": "http://pelias.mapzen.com",
    "textAnalyzer": "libpostal",
    "pipService": "http://localhost:3102",
    "services": {
      "placeholder": {
        "url": "http://localhost:3000/parser",
        "timeout": 100,
        "retries": 1
      }
    }
  }
```

Currently `timeout` and `retries` default to `250` and `3`, respectively.

**Comment by dianashk** [comment]
Outstanding issues based on acceptance test results on `dev`:

- [ ] Sorting of places using population/popularity and layer granularity
    - [/v1/search?text=california](http://pelias.github.io/compare/#/v1/search%3Ftext=california)
    - [/v1/search%3Fsize=50&text=wales](http://pelias.github.io/compare/#/v1/search%3Fsize=50&text=wales)
- [ ] When `layers` filter is provided do not show results from other layers
    - [/v1/search%3Flayers=county&text=bronx](http://pelias.github.io/compare/#/v1/search%3Flayers=county&text=bronx)
- [ ] Default english names should take into consideration the display names for places that need County or City in the name
    - [/v1/search%3Flayers=county&text=Lake, MT](http://pelias.github.io/compare/#/v1/search%3Flayers=county&text=Lake,%20MT)

**Pull request merged**